agent/src/com/cloud/agent/Agent.java

```diff
@@ -600,6 +608,47 @@ protected void processRequest(final Request request, final Link link) {
         }
     }
 
+    private Answer setupDirectDownloadCertificate(SetupDirectDownloadCertificate cmd) {
+        String certificate = cmd.getCertificate();
+        String certificateName = cmd.getCertificateName();
+        s_logger.info("Importing certificate " + certificateName + " into keystore");
+
+        final File agentFile = PropertiesUtil.findConfigFile("agent.properties");
+        if (agentFile == null) {
+            return new Answer(cmd, false, "Failed to find agent.properties file");
+        }
+
+        final String keyStoreFile = agentFile.getParent() + "/" + KeyStoreUtils.defaultKeystoreFile;
```

Comment:
The whole setup process could be implemented as a shell script with the java code calling that script to perform the job.

Reply:
I agree. You can do that, or you can do everything in Java (reading and writing files).

Reply:
Yes, I prefer using this java approach for those that can be done instead of using a script file. If you prefer I can refactor it to a script file

Reply:
I would prefer going the script way, perhaps re-using one of the import keystore/cert scripts. Alternatively, `KeyStore` could be read and manipulated (keys imported etc) by using Java only, agreeing with what @rafaelweingartner suggested.